File: py-jianzhi-round1/jianzhi56-II.py
class Solution:
    def singleNumber(self, nums) -> int:
        res = 0
        for i in range(32):
            cnt = 0  # 记录当前 bit 有多少个1
            bit = 1 << i  # 记录当前要操作的 bit
            for num in nums:
                if num & bit != 0:
                    cnt += 1
            if cnt % 3 != 0:
                # 不等于0说明唯一出现的数字在这个 bit 上是1
                res |= bit

        return res - 2 ** 32 if res > 2 ** 31 - 1 else res


## https://www.pythonheidong.com/blog/article/251304/
## https://leetcode-cn.com/problems/shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-ii-lcof/solution/mian-shi-ti-56-ii-shu-zu-zhong-shu-zi-chu-xian-d-4/
## 解释地很清楚. 思路可以说是非常明显了. 关键就是求解了.
## 第二个链接给的求解方法比较高端, 第一个链接给的方法就比较直白, 但是肯定不是最好的.

File: JavaProj/src/Jianzhi56II.java
public class Jianzhi56II {
    public int singleNumber(int[] nums) {
        // 位运算.
        // 遍历32位里面的每一位.
        // 在第i位, nums里面所有数字在这个位上是1的数字的个数, 看该数能被3整除.
        // 如果某一个位上1的个数不能被3整除, 那么所求的数在这个位上是1, 也就是因为它有个1, 导致的不能被3整除;
        int res = 0;
        for (int i = 1, mask = 1; i <= 32; i++, mask <<= 1){
            int cnt = 0;
            for (int num : nums){
                if ((num & mask) != 0){
                    cnt++;
                }
            }
            if (cnt % 3 != 0){
                res |= mask;
            }
        }
        return res;
    }
    public static void main(String []args){
        Jianzhi56II s = new Jianzhi56II();
        int [] res = new int[]{3,4,3,3};
        System.out.print(s.singleNumber(res));
    }
}
